aten/src/ATen/core/dispatch/DispatchTable.h

```diff
@@ -113,33 +113,55 @@ class KernelTable_ final {
 class DispatchTable final {
  public:
   DispatchTable(const FunctionSchema& schema)
-  : kernels_()
+  : kernels_(make_left<detail::KernelTable_, DispatchTableEntry>())
   , dispatch_strategy_(get_dispatch_strategy_(schema))
   , operator_name_(schema.name()) {}
 
   /**
    * Register a kernel in the table at some dispatch key.
    * @param dispatch_key Dispatch key to define when this kernel is selected.
-   *        If this is TensorTypeIds::undefined(), this registers a fallback
-   *        kernel that is called whenever no other kernel matches.
    * @param kernel Concrete kernel function implementation to register
    */
   void setKernel(
       TensorTypeId dispatch_key,
       const DispatchTableEntry& kernel) {
-    const bool is_fallback_kernel = (dispatch_key == TensorTypeIds::undefined());
-    AT_ASSERTM(is_fallback_kernel || dispatch_strategy_.is_valid_, "Tried to register a kernel with a dispatch key for operator schema ", operator_name_, " that doesn't have tensor arguments.");
-    kernels_.set(dispatch_key, kernel, operator_name_);
+    AT_ASSERTM(dispatch_key != TensorTypeIds::undefined());
+    AT_ASSERTM(dispatch_strategy_.is_valid_, "Tried to register a kernel with dispatch key ", detail::dispatch_key_to_string(dispatch_key), " for operator ", operator_name_, " that doesn't have tensor arguments.");
+    AT_ASSERTM(kernels_.is_left(), "Tried to register a kernel with dispatch key ", detail::dispatch_key_to_string(dispatch_key)," for operator ", operator_name_, ", which already has a catch-all kernel registered. An operator can only have either a catch-all kernel or kernels with dispatch keys.");
+    kernels_.left().set(dispatch_key, kernel, operator_name_);
   }
 
   /**
    * Deregister the kernel for some dispatch key.
    *
-   * @param dispatch_key Dispatch key to unregister. If this is
-   *        TensorTypeIds::undefined(), it will deregister the fallback kernel.
+   * @param dispatch_key Dispatch key to unregister.
    */
   void removeKernelIfExists(TensorTypeId dispatch_key) {
-    kernels_.removeIfExists(dispatch_key, operator_name_);
+    AT_ASSERTM(kernels_.is_left(), "Tried to remove the kernel for dispatch key ", detail::dispatch_key_to_string(dispatch_key), " for operator ", operator_name_, ", which only has a catch-all kernel.");
+    kernels_.left().removeIfExists(dispatch_key, operator_name_);
+  }
+
+  /**
+   * Register a catch-all kernel that is called for this operator
+   * independent of the inputs. An operator can have either
+   * a catch-all kernel or a set of kernels with concrete
+   * dispatch keys, not both.
+   */
+  void setCatchallKernel(const DispatchTableEntry& kernel) {
+    if (kernels_.is_right()) {
+      AT_WARN("Registered a catch-all kernel for operator ", operator_name_," that overwrote a previously registered catch-all kernel for the same operator.");
```

Comment:
why do we allow to override the catch all kernel? why not error here?

this break symmetry with register/unregister calls (i.e. the second unregister is going to fail)

Reply:
We need this for the Jupyter use case unfortunately. Re-executing the same cell should not hard fail. To be more concrete:

    // this will fail
    static auto registry = torch::RegisterOperators()
      .op("my::op", torch::RegisterOperators::options()
        .kernel<CatchallKernel>()
        .kernel<CatchallKernel>());

    // this will not fail but show a warning
    static auto registry1 = torch::RegisterOperators()
      .op("my::op", torch::RegisterOperators::options()
        .kernel<CatchallKernel>());
    static auto registry2 = torch::RegisterOperators()
      .op("my::op", torch::RegisterOperators::options()
        .kernel<CatchallKernel>());

The second unregister is not going to fail, we keep a list of all registered catch-all kernels and there's a test case for that: https://github.com/pytorch/pytorch/pull/20773/files#diff-e0d1d9f57915623b4a281362d41aa932R187